keps/sig-network/2943-networkpolicy-status/README.md

```diff
@@ -0,0 +1,641 @@
+feedback is provided, so Mary needs the cluster to tell her what's wrong with her policy.
+
+#### Story 2
+Samantha, the cluster admin of a Kubernetes cluster is watching the complaints about Network 
+Policies not working in the cluster raise after she upgraded the cluster CNI to the latest 
+version.
+
+Users decided to create Network Policies with some fancy new features that were announced 
+in the release notes of a random CNI, not knowing that the CNI used in Samantha's cluster 
+is different and still cannot implement the new fields.
+
+Samantha needs to know exactly what are the complaints and why the CNI cannot parse the 
+new rules, so she can explain to the users and also plan their monthly cluster upgrade to 
+add the required features.
+
+
+### Notes/Constraints/Caveats (Optional)
+The implementation of a status field depends on the CNI to provide this feedback. It's 
+not up to Kubernetes to know what happened, but to the controller of that object.
+
+If decided to use a "generic" approach of "controller / status / description" we cannot 
+guarantee that the CNI providers will follow the same standard/defaulting of the messages, 
+and this might be frustrating for users.
+
+### Risks and Mitigations
+Wrong RBAC in .status subresource field may lead to someone adding wrong status to users, and this
+may get mitigated by CNIs when using the right RBAC that allows only their "controllers" to change 
+this subresource.
+
+## Design Details
+API changes to NetworkPolicy:
+* Add the `status` inside `NetworkPolicy` struct as following:
+```go
+// NetworkPolicy describes what network traffic is allowed for a set of Pods
+type NetworkPolicy struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Specification of the desired behavior for this NetworkPolicy.
+	// +optional
+	Spec NetworkPolicySpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+  // Status is the current state of the NetworkPolicy.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	// +optional
+	Status NetworkPolicyStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+```
+
+* Add the `NetworkPolicyStatus` struct as following:
+```go
+// NetworkPolicyStatus describe the current state of the NetworkPolicy.
+type NetworkPolicyStatus struct {
+	PolicyStatus []CNIPolicyStatus
+}
+```
+
+* Add CNIPolicyStatus type struct as the following:
+
+```go
+type CNIPolicyStatus struct {
+  // CNI defines the CNI name
+	CNI string
+  
+  // Status defines the status string (parsed, error, etc)
+  Status string
+
+  // Description defines the description of the Status (why it's on that status)
+  Description string
+}
```

Comment:
I'd recommend this use a metav1.Conditions array instead of just `status` and `description` fields, because:
- Conditions have the ObservedGeneration and LastTransitionTime fields that are essential for checking that the version of the status you're seeing matches the version of the object (that is, that things are not still being updated).
- Lets there be a an expected main summary condition like `Accepted` or `Valid` that, when its `status` is true, indicates that everything was parsed correctly. When the `status` is false, then the Reason and Message field can be used to indicate the exact problem (like I would imagine Description is used here).
- This also lets extra Conditions be easily added to the spec later (I would guess there might also be a use for a desired-true `Ready` Condition that lets you know when the rules have been completely programmed into the underlying CNI.

I asked the question above, but in the case that the status *does* need to be namespaced by CNI name, then in that case, this would look like:
```suggestion
type CNIPolicyStatus struct {
  // CNI defines the CNI name that this status is written by.
  CNI string
  
  // Conditions holds an array of metav1,Conditions that describe the state of the NetworkPolicy.
  // When a CNI or other NetworkPolicy controller implementation parses this object,
  // it must ensure that the `Accepted` Condition is updated with the parsing outcome.
  // The Reason and Description must indicate the cause of the failure, if there was one.
  
  // (the Conditions field should also have a default that adds the `Accepted` Condition
  // with a `status` of `Unknown`.)
 Conditions []metav1.Conditon
}
```

Ideally, the update will also include Reason constants for the common Condition reasons (like `EverythingOK`, `FailedParsing` or something).

The idea here is that this is a spec that needs to be implemented by lots of CNI plugins, so the more clear the behavior is, the less chance there is of drift between implementations.

Reply:
I'm actually a -1 on making custom Condition types, I think unless you're adding something substantial to the metav1.Condition type, it's much better to use it directly. In particular, the `ObservedGeneration` field is absolutely vital for ensuring that the Condition isn't stale.  Regardless, they'll deserialize the same, but code using the field defs directly will find it easier to use the metav1.Condition than a custom type. (I recommend this because I did it in Contour's HTTPProxy, and it was a mistake.)

I'd encourage supplying string constants for the Type, Reason and Message fields instead.

Reply:
I'm a -1 on custom Condition Types as well. I've updated the KEP to contain Conditions, I think we are on the same page here in this case

Reply:
Also, TIL that Conditions are actually a meta type. Let me fix here in the KEP